### DISCUSSION TITLE
Self-balancing Line Follower

### DISCUSSION THREAD

**Discussion by SaharshSS** [pybricks/support#683] [Q&A]
Hi again, I am building the self balancing line follower from your YouTube channel. I am using this code: https://github.com/pybricks/pybricks-projects/blob/master/sets/mindstorms-robot-inventor/other-models/balancer/main.py 
but with the technic hub and color distance sensor. For the motors I am using the angular motors from the Volvo articulated hauler (2x) and for the wheels (parts 56145 and 15413.) It works okay at first but falls over after a few seconds. How do I modify the code to fix this?

**Comment by dlech** [comment]
If you are using a different hub, then you probably need to tune the constants in

    duty = 0.018 * rate + 19 * angle + 0.45 * diff + 0.16 * speed

**Reply by SaharshSS** [reply]
Thanks.

**Reply by SaharshSS** [reply]
What specific values do I adjust for each motion i.e. if it overcompensates which value should I adjust?

**Reply by laurensvalk** [reply]
As a starting point, try reducing everything at once, like so:

```python
# Original
duty = 0.018 * rate + 19 * angle + 0.45 * diff + 0.16 * speed

# Scale it down, because your motor is much stronger
duty = duty * 0.5
```

Try not just 0.5, but try higher and lower values too to see what is about right right. The lower, the less aggressive and responsive.

Your build is quite similar to the one in our example, so the relative numbers should be about the same. But the motor you use is much more powerful, so that's why the total value needs to be scaled down a bit to produce the same behavior.

**Reply by SaharshSS** [reply]
Here is my code, no matter what I scale the duty by, it cannot balance, I also have a line with dark blue painters tape and the color distance sensor doesn't seem to recognize it at all.

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, ColorDistanceSensor
from pybricks.parameters import Port, Direction
from pybricks.tools import wait, StopWatch
from pybricks.geometry import Axis

# Initialize motors.
left = Motor(Port.C, Direction.COUNTERCLOCKWISE)
right = Motor(Port.D)
left.reset_angle(0)
right.reset_angle(0)

# Initialize hub and color sensor.
# You can also use the TechnicHub and/or the ColorDistanceSensor
# instead.
hub = TechnicHub()
sensor = ColorDistanceSensor(Port.A)

# Initialize position buffer for speed calculation.
DT = 2
window = int(300 / DT)
buf = [0] * window
idx = 0
angle = 0
DRIVE_SPEED = 100
PAUSE = 500

# Timer to generate reference position.
watch = StopWatch()

while True:
    # Get angular rate and estimated angle.
    rate = hub.imu.angular_velocity(Axis.Y)
    angle += rate * DT / 1000

    # Get motor position.
    position = (left.angle() + right.angle()) / 2

    # Calculate motor speed.
    speed = (position - buf[idx]) / (window * DT) * 1000
    buf[idx] = position
    idx = (idx + 1) % window

    # Calculate reference position, which just grows linearly with
    # time.
    reference = -max(watch.time() - PAUSE, 0) / 1000 * DRIVE_SPEED

    # Calculate duty cycle.
    diff = position - reference
    duty = 0.018 * rate + 19 * angle + 0.45 * diff + 0.16 * speed
    duty = duty * 0.7

    # Account for battery level and type.
    duty *= 7200 / hub.battery.voltage()

    # Calculate steering.
    reflection = sensor.reflection()
    steering = (reflection - 28) * 0.6

    # Apply duty cycle for balancing and steering.
    left.dc(duty + steering)
    right.dc(duty - steering)

    # Wait some time.
    wait(DT)
```

**Reply by laurensvalk** [reply]
To make things easier, start by balancing in place instead of following a line. To do that, set the position reference and steering to zero. Find where these variables are defined and replace them with these:

```python
DRIVE_SPEED = 0
```

```python
steering = 0
```

**Reply by SaharshSS** [reply]
Thank you, when in place and the duty is multiplied by 0.75, it balances okay, but the wobble always makes it fall over on one side or the other. I'll send a video in a bit.

**Comment by laurensvalk** [comment]
> It works okay at first but falls over after a few seconds. 

Is it stable at first? And then suddenly falls over, [like this](https://github.com/pybricks/support/issues/232)?

**Reply by SaharshSS** [reply]
No it violent rocks back and forth for a few seconds, barely staying upright.

**Comment by SaharshSS** [comment]
![Photo on 7-7-22 at 10 58 AM](https://user-images.githubusercontent.com/67810603/177838979-20fee684-d15a-4f01-9fe0-feb74f9c3cf3.jpg)
A photo of my build

**Comment by SaharshSS** [comment]
Would it be possible to find some documentation on everything in the duty line?

**Comment by Werner-G** [comment]
Hi SaharshSS, did you manage to get the line follower worked with the Technic Hub?
